lib/existence.ex

```diff
@@ -93,50 +131,276 @@ defmodule Existence do
+    end)
+
+    data = %{checks: checks, ets_tab: ets_tab}
+
+    case Keyword.get(args, :state, :error) do
+      :ok -> {:ok, :healthy, data}
+      _ -> {:ok, :unhealthy, data}
+    end
+  end
+
+  @impl true
+  def callback_mode(), do: [:state_functions, :state_enter]
+
+  @impl true
+  def terminate(_reason, _state, data), do: set_state(data, :terminate)
+
+  # ________unhealthy
+  @doc false
+  def unhealthy(:enter, state, data) when state in [:healthy, :unhealthy] do
+    set_state(data, :unhealthy)
+    :keep_state_and_data
+  end
+
+  def unhealthy(:info, {:check_result, result, check_id}, data) do
+    set_check_state(data, check_id, result)
+
+    if is_ets_healthy?(data),
+      do: {:next_state, :healthy, data},
+      else: {:keep_state, data}
+  end
+
+  def unhealthy(:info, {:DOWN, ref, :process, pid, :normal}, data) do
+    find_check(pid, ref, data)
+    |> maybe_respawn_check()
 
+    :keep_state_and_data
+  end
+
+  def unhealthy(:info, {:DOWN, ref, :process, pid, error}, data) do
+    {check_id, _check_params} = check = find_check(pid, ref, data)
+    maybe_respawn_check(check)
+    set_check_state(data, check_id, error)
+
+    :keep_state_and_data
+  end
+
+  def unhealthy(:info, {:spawn_check, check_id}, data),
+    do: {:keep_state, spawn_check(check_id, data)}
+
+  # ________healthy
   @doc false
-  defdelegate child_spec(init_arg), to: GenCheck
+  def healthy(:enter, state, data) when state in [:healthy, :unhealthy] do
+    set_state(data, :healthy)
+    :keep_state_and_data
+  end
+
+  def healthy(:info, {:check_result, result, check_id}, data) do
+    set_check_state(data, check_id, result)
+
+    case result do
+      :ok -> {:keep_state, data}
+      _err -> {:next_state, :unhealthy, data}
+    end
+  end
+
+  def healthy(:info, {:DOWN, ref, :process, pid, :normal}, data) do
+    find_check(pid, ref, data)
+    |> maybe_respawn_check()
+
+    :keep_state_and_data
+  end
+
+  def healthy(:info, {:DOWN, ref, :process, pid, error}, data) do
+    {check_id, _check_params} = check = find_check(pid, ref, data)
+    maybe_respawn_check(check)
+    set_check_state(data, check_id, error)
+
+    {:next_state, :unhealthy, data}
+  end
+
+  def healthy(:info, {:spawn_check, check_id}, data),
+    do: {:keep_state, spawn_check(check_id, data)}
+
+  # ________helpers
+  defp ets_table_name(name), do: Module.concat(name, Table)
+
+  defp find_check(pid, ref, %{checks: checks}),
+    do: Enum.find(checks, nil, fn {_check_id, params} -> {pid, ref} == params.spawn_proc end)
+
+  defp maybe_respawn_check({check_id, check_params}),
+    do: Process.send_after(self(), {:spawn_check, check_id}, check_params.interval)
+
+  defp maybe_respawn_check(_invalid_check), do: :ok
+
+  defp spawn_check(check_id, %{checks: checks} = data) do
+    %{mfa: {m, f, a}, timeout: timeout} = params = Keyword.fetch!(checks, check_id)
+    from = self()
+
+    {pid, ref} =
```

Comment:
I couldn't quite figure out exactly what the reference monitor is used for. 🤔

Reply:
Erlang reference for a monitor is described in a [processes documentation](https://www.erlang.org/doc/reference_manual/processes.html#monitors).

In this particular case `MonitorRef` is used to uniquely identify spawned processes/checks.

When process is terminated, `{'DOWN', Ref, process, Pid2, Reason}` message is send to `:gen_statem` and `Ref` is used to uniquely identify given check using `find_check/3` and later take some further actions on that check, for example:
```elixir
  def healthy(:info, {:DOWN, ref, :process, pid, error}, data) do
    {check_id, _check_params} = check = find_check(pid, ref, data)
    maybe_respawn_check(check)
    set_check_state(data, check_id, error)

    {:next_state, :unhealthy, data}
  end
```